### ISSUE TITLE
Improve the state.js contract error handling logic

### ISSUE THREAD

**Issue by taoeffect** [okTurtles/group-income#1165]
### Problem

At the moment, the way state.js works for contract errors (with the automated banning system) has a few issues:

- It's kinda whacky
- It's extraordinarly contract-specific: it assumes we're dealing with identities (the identity contract) whom we can ban from a group (the group contract)
- It kinda breaks sometimes

### Solution

- We must separate the contract error handling into a framework-level and an application-level
- We need to handle contract errors in a more generic way while still allowing the app itself to handle them in a more specific way (which includes the possibility of automated bans)
- We need to make sure that messages that lead to contract errors are not discarded and re-processed, but ignored
- We need to still allow application-specific error handling for errors
- We need to allow developers to create and invent their own error classes for exceptions and not rely on any sort of specific error classes - leaving that to the application-specific logic

Note: instead of complicated auto-banning logic, we could simply ignore those messages, and post a notification to the UI stating which user's client sent a bad message, and what they were trying to do, and leave it up to the group to deal with banning them.

### Additional considerations and examples

- Users might have situations where they modify a contract to remove a no longer needed action. For example, we did this with `'gi.contracts/group/resetMonth'`. This resulted in the entire contract being unusable because:

    ```
    processMutation: error Error Error: chelonia: action not whitelisted: 'gi.contracts/group/resetMonth'
    [ERROR] exception GIErrorDropAndReprocess in handleEvent while processing <op_ae|21XWnNM8SQzDsW6gMZLwhVfd8i2SXRc6rXHAzmdhSMXjDGrfj1 of 21XWnNP6RkUF7sXAUZ5pZE9PgN7RnXdwo5M1yyjMKXrmdRovW1>! Error during processMutation: chelonia: action not whitelisted: 'gi.contracts/group/resetMonth' GIErrorDropAndReprocess: Error during processMutation: chelonia: action not whitelisted: 'gi.contracts/group/resetMonth'
    reverting to previous state! 
    [ERROR] exception GIErrorDropAndReprocess in handleEvent while processing <op_ae|21XWnNW2vZzVq2ekurqehMKNdcVMQ5bEnu9841ocAjzHJeqoH5 of 21XWnNP6RkUF7sXAUZ5pZE9PgN7RnXdwo5M1yyjMKXrmdRovW1>! ignoring message until page refresh: <op_ae|21XWnNW2vZzVq2ekurqehMKNdcVMQ5bEnu9841ocAjzHJeqoH5 of 21XWnNP6RkUF7sXAUZ5pZE9PgN7RnXdwo5M1yyjMKXrmdRovW1> GIErrorDropAndReprocess: ignoring message until page refresh: <op_ae|21XWnNW2vZzVq2ekurqehMKNdcVMQ5bEnu9841ocAjzHJeqoH5 of 21XWnNP6RkUF7sXAUZ5pZE9PgN7RnXdwo5M1yyjMKXrmdRovW1>
    reverting to previous state! 
    [ERROR] exception GIErrorDropAndReprocess in handleEvent while processing <op_ae|21XWnNM6Df5tWeBJvNB9emnqMd5jeuyiXuFk5BuX4M7QHUCWDa of 21XWnNP6RkUF7sXAUZ5pZE9PgN7RnXdwo5M1yyjMKXrmdRovW1>! ignoring message until page refresh: <op_ae|21XWnNM6Df5tWeBJvNB9emnqMd5jeuyiXuFk5BuX4M7QHUCWDa of 21XWnNP6RkUF7sXAUZ5pZE9PgN7RnXdwo5M1yyjMKXrmdRovW1> GIErrorDropAndReprocess: ignoring message until page refresh: <op_ae|21XWnNM6Df5tWeBJvNB9emnqMd5jeuyiXuFk5BuX4M7QHUCWDa of 21XWnNP6RkUF7sXAUZ5pZE9PgN7RnXdwo5M1yyjMKXrmdRovW1>
    ...etc...
    ```

- See also issue #1176 for a different error
- Sometimes we get a `[CRITICAL ERROR] NOT EXPECTING EVENT!` because things happen too fast - removing a contract while still having not processed some messages for it, as for example [here](https://github.com/okTurtles/group-income/pull/1143/files#diff-0e4600578c772d3d71d59082adec0332af88c00c5dacdb5d9ed97e5cdf40fc28R88) and [here](https://okturtles.slack.com/archives/C0EH7P20Y/p1649298837056349).

This issue also depends on #1214 being closed:

> [Slack] Greg: the issue is that right now the data and computation are separated, which means for example, if the rule is that if exceptions cause events to be ignored and skipped over (which likely seems to be the case for #1165), then once the issue in the contract has been fixed to prevent the exception, anyone who joins afterward and resyncs the old event (that existing clients skipped), they won’t encounter the exception, they will process the message, and therefore will have a different state from the other clients. This is a huge no-no, so I need to close #1214 along with #1165, so that each event specifies the exact code it uses to process itself
